src/libguac/client.c

```diff
@@ -701,6 +739,120 @@ int guac_client_owner_supports_required(guac_client* client) {
     
 }
 
+/**
+ * A callback function that is invokved by guac_client_owner_notify_join() to
+ * notify the owner of a connection that another user has joined the
+ * connection, returning zero if the message is sent successfully, or non-zero
+ * if an error occurs.
+ *
+ * @param user
+ *     The user to send the notification to, which will be the owner of the
+ *     connection.
+ *
+ * @param data
+ *     The data provided to the callback, which is the user that is joining the
+ *     connection.
+ *
+ * @return
+ *     Zero if the message is sent successfully to the owner, otherwise
+ *     non-zero, cast as a void*.
+ */
+static void* guac_client_owner_notify_join_callback(guac_user* user, void* data) {
+
+    const guac_user* joiner = (const guac_user *) data;
+    int retval = 0;
+
+    char* owner = "owner";
+    if (user->info.username != NULL)
+        owner = strdup(user->info.username);
+
+    char* joinName = "anonymous";
+    if (joiner->info.username != NULL)
+        joinName = strdup(joiner->info.username);
+
+    guac_user_log(user, GUAC_LOG_DEBUG, "Notifying owner %s of %s joining.", owner, joinName);
+    
+    /* Send required parameters to owner. */
+    if (user != NULL) {
+        const char* args[] = { (const char*)joinName, NULL };
+        retval = guac_protocol_send_msg(user->socket, GUAC_MSG_CLIENT_JOINED, args);
+    }
+
+    else
+        retval = -1;
+
+    free(owner);
```

Comment:
Be careful here - you may be trying to free string constants here.

Reply:
Okay, I think I've managed to fix this. Maybe there's a more elegant way....

Reply:
Actually, do you even need to free these at all? I mean, you're making copies of fields from `info`, or copies of string constants. Why not just avoid making copies at all?

Reply:
That's a good question - I'm not sure why I made copies in the first place. Probably should just use the originals...

Reply:
Oh, I see why I did it - I was trying to come up with a good way to insert some other text in there in the case where `info.username` isn't available. Is there a better way to go about substituting in a string constant if the `username` field isn't populated, that I'm just not thinking about right now?

Reply:
Wouldn't it work if you just removed all the `strdup()` and `free()` calls?

Reply:
The string constant is statically allocated anyway, so you don't need to worry about cleaning it up. And the fields on `info` are (I hope) getting cleaned up elsewhere.

Reply:
Ah, okay - I think I've got it, now - hopefully that works better...